bitcoin/script/miniscript.h

```diff
@@ -1014,224 +1014,384 @@ bool DecomposeScript(const CScript& script, std::vector<std::pair<opcodetype, st
+            }
+            // n: wrapper
+            else if (in[0].first == OP_0NOTEQUAL) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::ZeroNotEqual, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            // Thresh
+            else if (last - in >= 3 && in[0].first == OP_EQUAL && ParseScriptNumber(in[1], k)) {
+                if (k < 1) return {};
+                in += 2;
+                to_parse.emplace_back(DecodeContext::ThreshW, 0, k);
+            }
+            // a: wrapper
+            else if (in[0].first == OP_FROMALTSTACK) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::Alt, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            // OP_ENDIF can be WRAP_J, WRAP_D, ANDOR, OR_C, OR_D, or OR_I
+            else if (in[0].first == OP_ENDIF) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::EndIf, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeAndV, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            // and_b
+            else if (in[0].first == OP_BOOLAND) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::AndB, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            // or_b
+            else if (in[0].first == OP_BOOLOR) {
+                ++in;
+                to_parse.emplace_back(DecodeContext::OrB, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+                to_parse.emplace_back(DecodeContext::MaybeSwap, -1, -1);
+                to_parse.emplace_back(DecodeContext::Expression, -1, -1);
+            }
+            else {
+                // Unrecognised expression
+                return {};
             }
-            return {};
+            break;
         }
-        if (have_else && negated) {
-            auto sub3 = DecodeSingle<Key>(in, last, ctx);
-            if (!sub3) return {};
-            return MakeNodeRef<Key>(NodeType::ANDOR, Vector(std::move(sub3), std::move(sub1), std::move(sub2)));
+        case DecodeContext::MaybeAndV: {
+            // If we reach a potential AND_V top-level, check if the next part of the script could be another AND_V child
+            if (in < last && in[0].first != OP_IF && in[0].first != OP_ELSE && in[0].first != OP_NOTIF && in[0].first != OP_TOALTSTACK && in[0].first != OP_SWAP) {
```

Comment:
The above-listed opcodes cannot end any miniscript. Therefore they cannot be used in and_v because [X] in and_v(X, Y) is a well-formed miniscript.